### PR TITLE
adapter: split out a new timestamp-oracle crate

### PR DESCRIPTION
**Pull request by danhhz**
Besides the legacy impl of the trait (which is left in place) this is nicely self-contained, so break it out into its own crate. The immediate motivation is so I can depend on the trait (and maybe the postgres impl) in persistcli without pulling in adapter.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

The first commit is just some `git mv`s, feel free to skim. All the interesting bits are in the second commit.

Looks like it's much easier to leave the Catalog impl in mz-adapter (seems fine), but that means we have a make a few more things `pub` that were previously `pub(crate)`: the constructors, a couple things on metrics, and the `mod tests` stuff.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by aljoscha — APPROVED** [review]
It's fine for me to move to it's own crate (I've meant to do that myself but haven't gotten round to it... :sweat_smile:). For naming, I'd probably use `mz_timestamp_oracle`, which is longer than `mz_tso`, but I had to think a second when reading `tso`.

**Comment by danhhz** [comment]
Okay, this is ready to go!

> It's fine for me to move to it's own crate (I've meant to do that myself but haven't gotten round to it... 😅). For naming, I'd probably use `mz_timestamp_oracle`, which is longer than `mz_tso`, but I had to think a second when reading `tso`.

I think my own math leans firmly the other way, it's easy to pick up jargon, but annoying to type long strings over and over (I got sick of typing "timestamp-oracle" just in updated this PR 😛), but I defer to you here. Done.

**Comment by danhhz** [comment]
TFTR!

**Review by ParkMyCar — APPROVED** [review]

File: src/timestamp-oracle/src/lib.rs
Comment (ParkMyCar) [inline src/timestamp-oracle/src/lib.rs:10]: Just a heads up that this might conflict with https://github.com/MaterializeInc/materialize/pull/23725, depending on which lands first. Definitely no need to wait for that PR to merge though, I can rebase on `main`!

Comment (danhhz) [reply src/timestamp-oracle/src/lib.rs:10]: Yup, I was about to give you the same heads up! This is set to merge on green, so hopefully it lands imminently 🤞

Comment (ParkMyCar) [reply src/timestamp-oracle/src/lib.rs]: Just merged the PR, looks like you rebased on `main`, hopefully it wasn't too bad 🙈

**Pull request merged**
